frame/system/src/lib.rs

```diff
@@ -485,19 +485,35 @@ decl_module! {
 		}
 
 		/// Make some on-chain remark.
+		/// 
+		/// # <weight>
+		/// - `O(1)`
+		/// # </weight>
 		#[weight = SimpleDispatchInfo::FixedNormal(MINIMUM_WEIGHT)]
 		fn remark(origin, _remark: Vec<u8>) {
 			ensure_signed(origin)?;
 		}
 
 		/// Set the number of pages in the WebAssembly environment's heap.
+		/// 
+		/// # <weight>
+		/// - `O(1)`
+		/// - 1 storage write.
+		/// # </weight>
 		#[weight = SimpleDispatchInfo::FixedOperational(MINIMUM_WEIGHT)]
 		fn set_heap_pages(origin, pages: u64) {
 			ensure_root(origin)?;
 			storage::unhashed::put_raw(well_known_keys::HEAP_PAGES, &pages.encode());
 		}
 
 		/// Set the new runtime code.
+		/// 
+		/// # <weight>
+		/// - `O(C + S)` where `C` length of `code` and `S` complexity of `can_set_code`
+		/// - 1 storage write (codec `O(C)`).
+		/// - 1 call to `can_set_code`: `O(S)` (calls `sp_io::misc::runtime_version` which is expensive).
+		/// - 1 event.
```

Comment:
I'm not sure how should we express this:

shouldn't it be: `complexity = O(C) + S with S cost of sp_io::misc::runtime_version` ?

Here S is not a variable complexity it is a constant one.

Reply:
Because [big O notation](https://en.wikipedia.org/wiki/Big_O_notation) only gives *an* upper bound and not *the* tight upper bound, both `O(C + S)` and `O(C) + S` are totally admissable. The only guarantee is "will never be higher than the complexity of a function in this category", so in principle it's fine either way.

I guess in practice we often try to give a reasonably tight upper bound, but I don't think we need to spend a lot of time discussing how tight exactly it should be. :shrug:

Reply:
The nice thing of `O(S)` (instead of `S`) is that there might be some constant costs associated with calling `can_set_code` (or as in your comment `runtime_version`) that are included with big O, but not if you just put it there.

Reply:
I think `O(C + S)` if S is constant complexity is equivalent to `O(C)` and so we should prefer the second one. By equivalent I mean that there is mathematically no differences at all (but I actually forgot a bit of my approximation math courses)

In the notion https://www.notion.so/Benchmarking-Philosophy-c5aafd34578f4be9ab8c8d7510e98314
It seems the strategy is to then come up with real number using benchmark. but S here is not varying so we don't benchmark its variation. so we will only benchmark with C varying and we will compute `w = K1 + K2*C` and so S will disappear in K1

> The nice thing of O(S) (instead of S) is that there might be some constant costs associated with calling can_set_code (or as in your comment runtime_version) that are included with big O, but not if you just put it there.

I don't understand this for me S is constant cost so when a chain will benchmark its stuff it will be included in K1 no matter what the chain is using under the hood

Reply:
PS: I'm not trying to be picky I'm just thinking that if we want to follow the strategy:
* for complexity: `O(M + logM + B + E)`
* benchmark will find constant for `weight(M, B, E) = K_1 + K_2 * M + K_3 * logM + B + E`

hmm actually I lost myself with this paragraph in notion:

> When doing empirical testing, we are unable to separate complexities which have the same order. This means that there could be many many more operations added to this function, of order O(M), O(logM), etc.. but it would not change our final formula as a function of M, B, and E:

I though benchmark will compute: `weight(M, B, E) = K_1 + K_2 * M + K_3 * logM + K_4*B + K_5*E`
So not sure anymore about how this works

Reply:
> I think O(C + S) if S is constant complexity is equivalent to O(C) and so we should prefer the second one. By equivalent I mean that there is mathematically no differences at all (but I actually forgot a bit of my approximation math courses)

The complexity of `can_set_code` (--> `S`) depends on the code size so `S` is not constant. And I actually don't know which of the two is greater (I actually guess that it is `S`) so I just added both of them.

Reply:
I meant constant costs *associated with*  `S`, not `S` having constant costs.

Reply:
I have not looked into the benchmarks much, yet, so I'm sad to tell you that I cannot be very helpful there. Maybe @shawntabrizi can weigh in .

Reply:
Ah I mixed myself indeed
`complexity(sp_io::misc::runtime_version(&code)) == O(code.len())`
so the cost of the extrinsic should be `O(C + C)= O(C)` no ?

At the end `weight(C) = K1 + K2*C` any constant cost in can_set_code wil be in K1 anyway.

Reply:
That's true, except that I don't know whether S might be `O(C^2)` or something even more crazy. I went the conservative way because I saw the note on `sp_io::misc::runtime_version` being expensive and relying on executing the code in question.

Reply:
hmm ok I also mistaken on the complexity of `sp_io::misc::runtime_version` then I'm ok to have S as complexity of `sp_io::misc::runtime_version`, which is unknown to me.

I guess to come up to a correct formula we should say this weight full block ?

Reply:
Maybe? I'm out of my depth here.
@gavofyork or @shawntabrizi What do you think?